config/src/test/java/org/springframework/security/config/http/OAuth2LoginBeanDefinitionParserTests.java

```diff
@@ -0,0 +1,739 @@
+		assertThat(authorizedClientRepositoryFieldVal)
+				.isInstanceOf(AuthenticatedPrincipalOAuth2AuthorizedClientRepository.class);
+
+		Field authorizedClientServiceField = authorizedClientRepositoryFieldVal.getClass()
+				.getDeclaredField("authorizedClientService");
+		authorizedClientServiceField.setAccessible(true);
+		Object authorizedClientServiceFieldVal = authorizedClientServiceField.get(authorizedClientRepositoryFieldVal);
+		assertThat(authorizedClientServiceFieldVal).isInstanceOf(MockOAuth2AuthorizedClientService.class);
+		assertThat(authorizedClientServiceFieldVal).isSameAs(oauth2AuthorizedClientService);
+	}
+
+	@Test
+	public void customAuthorizationRequestRepository() throws Exception {
+		this.spring.configLocations(this.xml("WithCustomAuthorizationRequestRepository")).autowire();
+
+		assertThat(authorizationRequestRepository).isInstanceOf(MockOAuth2AuthorizationRequestRepository.class);
+
+		Field authorizationRequestRepositoryField = oauth2LoginAuthenticationFilter.getClass()
+				.getDeclaredField("authorizationRequestRepository");
+		authorizationRequestRepositoryField.setAccessible(true);
+		Object authorizationRequestRepositoryFieldVal = authorizationRequestRepositoryField
+				.get(oauth2LoginAuthenticationFilter);
+		assertThat(authorizationRequestRepositoryFieldVal).isInstanceOf(MockOAuth2AuthorizationRequestRepository.class);
+		assertThat(authorizationRequestRepositoryFieldVal).isSameAs(authorizationRequestRepository);
+	}
+
+	@Test
+	public void requestWithSingleClientRegistrationAndCustomSuccessAuthenticationHandler() throws Exception {
+		this.spring.configLocations(this.xml("SingleClientRegistration-WithCustomAuthenticationHandler")).autowire();
+
+		MockHttpSession session = new MockHttpSession();
+		MockHttpServletRequest request = new MockHttpServletRequest("GET", "");
+		request.setServletPath("/login/oauth2/code/google");
+		request.setSession(session);
+		MockHttpServletResponse response = new MockHttpServletResponse();
+
+		OAuth2AuthorizationRequest authRequest = createOAuth2AuthorizationRequest(
+				clientRegistrationRepository.findByRegistrationId("google-login"));
+		authorizationRequestRepository.saveAuthorizationRequest(authRequest, request, response);
+
+		MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
+		params.add("code", "code123");
+		params.add("state", "state123");
+		this.mvc.perform(get("/login/oauth2/code/google").params(params).session(session))
+				.andExpect(status().isIAmATeapot());
+
+		Authentication authentication = this.securityContextRepository
+				.loadContext(new HttpRequestResponseHolder(request, response)).getAuthentication();
+		assertThat(authentication.getAuthorities()).hasSize(1);
+		assertThat(authentication.getAuthorities()).first().isInstanceOf(OAuth2UserAuthority.class)
+				.hasToString("ROLE_USER");
+	}
+
+	@Test
+	public void requestWithSingleClientRegistrationAndCustomFailureAuthenticationHandler() throws Exception {
```

Comment:
Please update test to `verify()` that the mock `AuthenticationFailureHandler` is called. We can also rename to `requestWhenCustomAuthenticationFailureHandlerThenCalled`